### PR TITLE
docs(element-internals): document form-associated callbacks

### PR DESCRIPTION
**Pull request by alicewriteswrongs**
This adds documentation of the form-associated lifecycle callbacks, explaining what they do and providing some quick examples of how they can be used.

### CODE REVIEW

**Review by christian-bromann — APPROVED** [review]
LGTM 👍

File: docs/components/form-associated.md [hunk 1]
```diff
@@ -123,10 +123,183 @@ the `ElementInternals` API, including [setting the element's
...
+}
+```
+
+### `formStateRestoreCallback`
+
+This method will be called in the event that the browser automatically fills
+out your form element, an event that could take place in two different
+scenarios. The first is that the browser can restore the state of an element
+after navigating or restarting, and the second is that an input was made using a
+form autofilling feature.
```
Comment (christian-bromann) [inline docs/components/form-associated.md:230]: Maybe this resolves the cspell issue?
```suggestion
form auto-filling feature. 
```

**Review by tanner-reits — APPROVED** [review]
Friendly reminder that we'll wanna propagate these to the 4.8 docs as well (I think that's when the lifecycle methods were added) 🙂

File: docs/components/form-associated.md [hunk 2]
```diff
@@ -123,10 +123,183 @@ the `ElementInternals` API, including [setting the element's
...
+
+### `formStateRestoreCallback`
+
+This method will be called in the event that the browser automatically fills
+out your form element, an event that could take place in two different
+scenarios. The first is that the browser can restore the state of an element
+after navigating or restarting, and the second is that an input was made using a
+form autofilling feature. 
+
+In any case, in order to correctly reset itself your form-associated component
```
Comment (rwaskiewicz) [inline docs/components/form-associated.md:232]: Non blocking/tiny copy edit to help with readability - this seemed a little too wordy to me. Thoughts on something like this?
```suggestion
In order to correctly reset, itself your form-associated component
```

File: docs/components/form-associated.md [hunk 3]
```diff
@@ -123,10 +123,183 @@ the `ElementInternals` API, including [setting the element's
...
+disassociates the element from a form element. The function is called with the
+form element as an argument. This could be used to set an `ariaLabel` when the
+form is ready to use, like so:
+
+```tsx title='src/components/form-associated-cb.tsx'
+import { Component, h, AttachInternals } from '@stencil/core';
+
+@Component({
+  tag: 'form-associated',
+  formAssociated: true,
```
Comment (rwaskiewicz) [inline docs/components/form-associated.md:147]: Non blocking - The other examples here set `shadow: true,` here as well. Should we add it here for consistency? IDK if folks will try to find some meaning in its absence here or not

File: docs/components/form-associated.md [hunk 4]
```diff
@@ -123,10 +123,183 @@ the `ElementInternals` API, including [setting the element's
...
+
+This is called when the form is reset, and should be used to reset the
+form-associated component's internal state and validation. For example, you
+could do something like the following:
+
+```tsx title="src/components/form-reset-cb.tsx"
+import { Component, h, State } from '@stencil/core';
+
+@Component({
+  tag: 'form-disabled-cb',
```
Comment (rwaskiewicz) [inline docs/components/form-associated.md:203]: I think this is a copy/paste error:
```suggestion
  tag: 'form-reset-cb',
```

File: docs/components/form-associated.md [hunk 5]
```diff
@@ -123,10 +123,183 @@ the `ElementInternals` API, including [setting the element's
...
+```
+
+### `formResetCallback`
+
+This is called when the form is reset, and should be used to reset the
+form-associated component's internal state and validation. For example, you
+could do something like the following:
+
+```tsx title="src/components/form-reset-cb.tsx"
+import { Component, h, State } from '@stencil/core';
```
Comment (rwaskiewicz) [inline docs/components/form-associated.md:200]: Missing import will cause this example to fail:
```suggestion
import { AttachInternals, Component, h, State } from '@stencil/core';
```

File: docs/components/form-associated.md [hunk 6]
```diff
@@ -123,10 +123,183 @@ the `ElementInternals` API, including [setting the element's
...
+
+@Component({
+  tag: 'form-disabled-cb',
+  shadow: true,
+  formAssociated: true,
+})
+export class MyComponent {
+  @State() cssClass: string = "";
```
Comment (rwaskiewicz) [inline docs/components/form-associated.md:208]: I'm not sure I fully understand how `cssClass` here is expected to work in this example. I don't see a place where we set this to anything other than the empty string here, and use that value on the `class` for the `input` element. Can you please help me understand the role it plays in this example?

File: docs/components/form-associated.md [hunk 7]
```diff
@@ -123,10 +123,183 @@ the `ElementInternals` API, including [setting the element's
...
+[`setFormValue`](https://developer.mozilla.org/en-US/docs/Web/API/ElementInternals/setFormValue)
+method on `ElementInternals`  provides some support for this use-case, allowing
+you to submit both a _value_ and a _state_, where the _state_ is not added to
+the form data sent to the server but could be used for storing some
+client-specific state. For instance, a pseudocode sketch of a date picker
+component that correctly restores whether the 'week' or 'month' view is active
+could look like:
+
+```tsx title="src/components/fa-date-picker.tsx"
+import { Component, h, State } from '@stencil/core';
```
Comment (rwaskiewicz) [inline docs/components/form-associated.md:249]: Similar to above:
```suggestion
import { AttachInternals, Component, h, State } from '@stencil/core';
```

File: docs/components/form-associated.md [hunk 8]
```diff
@@ -123,10 +123,183 @@ the `ElementInternals` API, including [setting the element's
...
+    const [date, view] = state.split("#");
+    this.view = view;
+    this.setValue(date);
+  }
+
+  render() {
+    return <input
+      class="date-picker"
+      mode={this.view}
+      onChange={onInputChange}
```
Comment (rwaskiewicz) [inline docs/components/form-associated.md:284]:
```suggestion
      onChange={this.onInputChange}
```

File: docs/components/form-associated.md [hunk 9]
```diff
@@ -123,10 +123,183 @@ the `ElementInternals` API, including [setting the element's
...
+week or month view.
+
+The
+[`setFormValue`](https://developer.mozilla.org/en-US/docs/Web/API/ElementInternals/setFormValue)
+method on `ElementInternals`  provides some support for this use-case, allowing
+you to submit both a _value_ and a _state_, where the _state_ is not added to
+the form data sent to the server but could be used for storing some
+client-specific state. For instance, a pseudocode sketch of a date picker
+component that correctly restores whether the 'week' or 'month' view is active
+could look like:
```
Comment (rwaskiewicz) [inline docs/components/form-associated.md:246]: Can we make sure this is a runnable example please? I'm not married to it being a Date Picker, but this should compile in a Stencil Component Starter project for folks learning about these callbacks and how to use them in the context of Stencil.

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]

File: docs/components/form-associated.md (hunk 3 above)
Comment (alicewriteswrongs) [reply docs/components/form-associated.md:147]: hmmmmmm yeah I think for consistency I'll just remove it on all of them

File: docs/components/form-associated.md (hunk 6 above)
Comment (alicewriteswrongs) [reply docs/components/form-associated.md:208]: just a copy-paste error

File: docs/components/form-associated.md (hunk 9 above)
Comment (alicewriteswrongs) [reply docs/components/form-associated.md:246]: yeah, just adjusted the render to make it compile

File: docs/components/form-associated.md [hunk 10]
```diff
@@ -123,10 +123,177 @@ the `ElementInternals` API, including [setting the element's
...
+```
+
+### `formResetCallback`
+
+This is called when the form is reset, and should be used to reset the
+form-associated component's internal state and validation. For example, you
+could do something like the following:
+
+```tsx title="src/components/form-reset-cb.tsx"
+import { Component, h, State, AttachInternals } from '@stencil/core';
```
Comment (rwaskiewicz) [inline docs/components/form-associated.md:199]: Minor, `State` is no longer used
```suggestion
import { Component, h, AttachInternals } from '@stencil/core';
```

File: docs/components/form-associated.md [hunk 11]
```diff
@@ -123,10 +123,177 @@ the `ElementInternals` API, including [setting the element's
...
+export class MyDatePicker {
+  @State() value: string = "";
+  @State() view: "weeks" | "months" = "weeks";
+
+  @AttachInternals()
+  internals: ElementInternals;
+
+  onInputChange(e) {
+    const date = e.target.value;
+    this.setValue(date)
```
Comment (rwaskiewicz) [inline docs/components/form-associated.md:260]: Does `this` get set correctly here? When I run this example, I get an error in the console.


https://github.com/ionic-team/stencil-site/assets/1930213/c74b5dfd-fc3c-4746-b6a5-7f8f67d8cd48

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]

File: docs/components/form-associated.md (hunk 11 above)
Comment (alicewriteswrongs) [reply docs/components/form-associated.md:260]: oh I forgot you need the little `e => cb(e)` indirection, I'll add that

**Review by rwaskiewicz — APPROVED** [review]
LGTM! Just reiterating Tanner's ask to backport this to v4.8, but that's non-blocking
